pallets/creditcoin/src/lib.rs

```diff
@@ -540,11 +551,74 @@ pub mod pallet {
 	#[pallet::hooks]
 	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
 		fn on_initialize(block_number: T::BlockNumber) -> Weight {
-			log::debug!("Cleaning up expired entries");
+			log::debug!("Cleaning up expired entries: {block_number:?}");
+
+			let mut current =
+				CleanupState::<T>::get().unwrap_or_else(|| StorageCleanupState::new(block_number));
+
+			let limit = T::PerBlockCleanupLimit::get();
+			let mut ask_count = 0u32;
+			let mut bid_count = 0u32;
+			let mut offer_count = 0u32;
+			let mut total_count = 0u32;
+
+			loop {
+				let StorageCleanupState { ask_orders, bid_orders, offers } = current;
+				log::debug!("ask orders block: {:?}, cursor: {}; bid orders block: {:?}, cursor: {}; offers block: {:?}, cursor: {}",
+					ask_orders.on_block,ask_orders.cursor.is_some(), bid_orders.on_block, bid_orders.cursor.is_some(),offers.on_block, offers.cursor.is_some(),);
+				let ask_cleanup = (ask_orders.on_block <= block_number && total_count < limit)
+					.then(|| {
+						let ask_cleanup = AskOrders::<T>::clear_prefix(
+							ask_orders.on_block,
+							limit.saturating_sub(total_count),
```

Comment:
In cases where the total count of a single storage prefix is larger than the limit, wouldn't this result in no deletion taking place at all?